### ISSUE TITLE
[cfe] Soundness issue with missing caller-side check

### ISSUE THREAD

**Issue by eernstg** [dart-lang/sdk#46784]
Consider the following program:

```dart
class C<X> {
  void m<Y extends X>(Y y) {}
}

void main() {
  C<num> c = C<int>();
  void Function<Z extends num>(Z) f = c.m;
  print(f.runtimeType);
}
```

This program is accepted by `dart` (from commit a4afb06b215eb9b40cb9cb17a6c69d03e413d8a0), it runs without any dynamic errors, and it prints `'<Y0 extends int>(Y0) => void'`.

However, this is a soundness violation:  The variable `f` holds a value whose dynamic type is not a subtype of the declared type of the variable.

We do not specify how to deal with this situation, but in a number of similar situations it is currently handled by performing a caller-side check (so the object which is the value of `c.m` is obtained, then it is checked that this object has the required type, and a dynamic error occurs if that is not the case).

**Comment by stefantsov** [comment]
Currently waiting on another issue: https://github.com/dart-lang/sdk/issues/46816.

**Comment by stefantsov** [comment]
@eernstg It seems that the following language tests should be updated:

* tests/language/covariant/type_parameter_test.dart,
* tests/language_2/covariant/type_parameter_test.dart.

**Comment by eernstg** [comment]
> tests/language/covariant/type_parameter_test.dart,

Right, line 31, evaluation of `b.f2` should throw. (My desktop has just been taken offline, so I'll create a CL when I'm back from the weekend). Line 37 is still correct.

> tests/language_2/covariant/type_parameter_test.dart

For the legacy version of the test, I'd recommend that we keep the old semantics, as usual, in the interest of stability.

**Comment by a-siva** [comment]
Does pkg/vm/test/incremental_compiler_test also need to be updated, it appears so from the approval comment for this test failure. Have issues been filed in the co19 repository for updates needed to tests/language/covariant/type_parameter_test.dart and tests/language_2/covariant/type_parameter_test.dart. Also is there an issue for fixing pkg/vm/test/incremental_compiler_test

**Comment by stefantsov** [comment]
@a-siva Thank you for letting us know about the test failures!

> Does pkg/vm/test/incremental_compiler_test also need to be updated, it appears so from the approval comment for this test failure.

Interesting! I didn't approve the failure on pkg/vm/test/incremental_compiler_test, and I can see that on the 'Aggregated test results' page for my CL: https://dart-ci.firebaseapp.com/cl/208881/3. I'll investigate if my change affected the test and try to fix that.

> Have issues been filed in the co19 repository for updates needed to tests/language/covariant/type_parameter_test.dart and tests/language_2/covariant/type_parameter_test.dart. Also is there an issue for fixing pkg/vm/test/incremental_compiler_test

These two tests are discussed on this issue, in the comments above. One of them should be updated, and the failure on the other one should be fixed soon. The tests in the dart-lang/sdk repository, and I'll implement the fixes for both of them.

**Comment by eernstg** [comment]
[Tests](https://github.com/dart-lang/sdk/issues/46784#issuecomment-894268869) adjusted in https://dart-review.googlesource.com/c/sdk/+/209860.
